In [1]:
import pandas as pd
import numpy as np
import json
import seaborn as sns

In [104]:
data = [json.loads(line) for line in open('yelp_academic_dataset_business.json', 'r')]
#data = []
#for line in open('yelp_academic_dataset_business.json', 'r'):
#   data.append(json.loads(line))

In [105]:
from pandas import DataFrame
df = DataFrame(data)
df.head()

business_id                      name  \
0  f9NumwFMBDn751xgFiRbNA  The Range At Lake Norman   
1  Yzvjg0SayhoZgCljUJRF9Q         Carlos Santo, NMD   
2  XNoUzKckATkOD1hP6vghZg                   Felinus   
3  6OAZjbxqM5ol29BuHsil3w      Nevada House of Hose   
4  51M2Kk903DFYI6gnB5I6SQ   USE MY GUY SERVICES LLC   

                     address             city state postal_code   latitude  \
0            10913 Bailey Rd        Cornelius    NC       28031  35.462724   
1  8880 E Via Linda, Ste 107       Scottsdale    AZ       85258  33.569404   
2      3554 Rue Notre-Dame O         Montreal    QC     H4C 1P4  45.479984   
3             1015 Sharp Cir  North Las Vegas    NV       89030  36.219728   
4         4827 E Downing Cir             Mesa    AZ       85205  33.428065   

    longitude  stars  review_count  is_open  \
0  -80.852612    3.5            36        1   
1 -111.890264    5.0             4        1   
2  -73.580070    5.0             5        1   
3 -115.127725    2.5             3        0   
4 -111.726648    4.5            26        1   

                                          attributes  \
0  {'BusinessAcceptsCreditCards': 'True', 'BikePa...   
1  {'GoodForKids': 'True', 'ByAppointmentOnly': '...   
2                                               None   
3  {'BusinessAcceptsCreditCards': 'True', 'ByAppo...   
4  {'BusinessAcceptsCreditCards': 'True', 'ByAppo...   

                                          categories  \
0  Active Life, Gun/Rifle Ranges, Guns & Ammo, Sh...   
1  Health & Medical, Fitness & Instruction, Yoga,...   
2                   Pets, Pet Services, Pet Groomers   
3  Hardware Stores, Home Services, Building Suppl...   
4  Home Services, Plumbing, Electricians, Handyma...   

                                               hours  
0  {'Monday': '10:0-18:0', 'Tuesday': '11:0-20:0'...  
1                                               None  
2                                               None  
3  {'Monday': '7:0-16:0', 'Tuesday': '7:0-16:0', ...  
4  {'Monday': '0:0-0:0', 'Tuesday': '9:0-16:0', '...

In [106]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209393 entries, 0 to 209392
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   business_id   209393 non-null  object 
 1   name          209393 non-null  object 
 2   address       209393 non-null  object 
 3   city          209393 non-null  object 
 4   state         209393 non-null  object 
 5   postal_code   209393 non-null  object 
 6   latitude      209393 non-null  float64
 7   longitude     209393 non-null  float64
 8   stars         209393 non-null  float64
 9   review_count  209393 non-null  int64  
 10  is_open       209393 non-null  int64  
 11  attributes    180348 non-null  object 
 12  categories    208869 non-null  object 
 13  hours         164550 non-null  object 
dtypes: float64(3), int64(2), object(9)
memory usage: 22.4+ MB


In [107]:
df.drop(["business_id", "attributes", "hours"], axis=1, inplace=True)

In [108]:
df.describe()

latitude      longitude          stars   review_count  \
count  209393.000000  209393.000000  209393.000000  209393.000000   
mean       38.579934     -97.390217       3.538055      36.937505   
std         4.940448      16.718535       1.023543     123.343597   
min        21.497258    -158.025525       1.000000       3.000000   
25%        33.638658    -112.269476       3.000000       4.000000   
50%        36.147408    -111.743531       3.500000       9.000000   
75%        43.611693     -79.972679       4.500000      27.000000   
max        51.299943     -72.806550       5.000000   10129.000000   

             is_open  
count  209393.000000  
mean        0.806632  
std         0.394940  
min         0.000000  
25%         1.000000  
50%         1.000000  
75%         1.000000  
max         1.000000

In [109]:
df.describe(include=["object"])

name address       city   state postal_code  \
count      209393  209393     209393  209393      209393   
unique     157229  164423       1251      37       18605   
top     Starbucks          Las Vegas      AZ       89109   
freq         1184    8679      31631   60803        3388   

                        categories  
count                       208869  
unique                      102494  
top     Nail Salons, Beauty & Spas  
freq                          1055

In [110]:
df["city"].value_counts().to_frame()

city
Las Vegas                      31631
Toronto                        20366
Phoenix                        20171
Charlotte                      10422
Scottsdale                      9342
...                              ...
montreal                           1
Regional Municipality of York      1
Elrama                             1
North Phoenix                      1
Sainte-thérèse                     1

[1251 rows x 1 columns]

In [111]:
df1=df.loc[df["city"]=="Toronto"]
df1.shape

(20366, 11)

In [112]:
df1.is_open.value_counts()

1    14962
0     5404
Name: is_open, dtype: int64

In [113]:
df1=df1.loc[df1["is_open"]==1]

In [114]:
df1.describe(include=["object"])

name address     city  state postal_code          categories
count       14962   14962    14962  14962       14962               14928
unique      12464   12827        1      3        5136                9369
top     Starbucks          Toronto     ON              Food, Coffee & Tea
freq          130     283    14962  14960         108                 133

In [115]:
# Create one row for each series that contain comma-separated items
df2 = df1.assign(categories = df1.categories.str.split(', ')).explode('categories')
df2.shape


(59696, 11)

In [116]:
total=len(df2["categories"].value_counts())
print(f"total categories in the data set: {total}")
df2["categories"].value_counts().to_frame()[:10]


total categories in the data set: 928


categories
Restaurants                      5462
Food                             3240
Shopping                         2656
Beauty & Spas                    1675
Nightlife                        1179
Bars                             1056
Coffee & Tea                     1055
Health & Medical                 1052
Local Services                    910
Event Planning & Services         784

In [117]:
missing_data = df1.isnull()
for i in list(missing_data):
    print (i)
    print (missing_data[i].value_counts())
    print ("")
    

name
False    14962
Name: name, dtype: int64

address
False    14962
Name: address, dtype: int64

city
False    14962
Name: city, dtype: int64

state
False    14962
Name: state, dtype: int64

postal_code
False    14962
Name: postal_code, dtype: int64

latitude
False    14962
Name: latitude, dtype: int64

longitude
False    14962
Name: longitude, dtype: int64

stars
False    14962
Name: stars, dtype: int64

review_count
False    14962
Name: review_count, dtype: int64

is_open
False    14962
Name: is_open, dtype: int64

categories
False    14928
True        34
Name: categories, dtype: int64



we have 928 different categories and among data 34 of them are missing. it's take a lot of time to find the appropriate categories for them so we're going to drop them. 

In [118]:
df1.dropna(subset=["categories"], axis=0, inplace=True)

### time to change data type

In [124]:
df1[["name"]]=df1[["name"]].astype("str")
df1[["address"]]=df1[["address"]].astype("str")
df1[["city"]]=df1[["city"]].astype("str")
df1[["state"]]=df1[["state"]].astype("str")
#postal code in canada are a combinatian of numbers and string
df1[["postal_code"]]=df1[["postal_code"]].astype("str")
df1[["stars"]]=df1[["stars"]].astype("float")
df1[["categories"]]=df1[["categories"]].astype("str")
df1.dtypes


name             object
address          object
city             object
state            object
postal_code      object
latitude        float64
longitude       float64
stars           float64
review_count      int64
is_open           int64
categories       object
dtype: object

Now we remove any whitespace in names,address,city, and state and also change all letters to lowercase

In [120]:
df1.name=df1.name.str.strip().str.lower()
df1.address=df1.address.str.strip().str.lower()
df1.city=df1.city.str.strip().str.lower()
df1.state=df1.state.str.strip().str.lower()
df1.categories=df1.categories.str.strip().str.lower()

In [121]:
df1.head()

name  \
9                                      xtreme couture   
26                                   air jordan store   
52                                bath and body works   
73                                   mi mi restaurant   
74  equipment sales and long term rentals - redlin...   

                                 address     city state postal_code  \
9           700 kipling avenue etobicoke  toronto    on     M8Z 5G3   
26                      306 yonge street  toronto    on     M5B 1R4   
52                  1900 eglintion ave e  toronto    on     M1L 2L9   
73                  688 gerrard street e  toronto    on     M4M 1Y3   
74  510 coronation dr. unit #18, unit 18  toronto    on     M1E 4X6   

     latitude  longitude  stars  review_count  is_open  \
9   43.624539 -79.529108    3.0            16        1   
26  43.656542 -79.381308    4.0             9        1   
52  43.727189 -79.293008    3.5             7        1   
73  43.666376 -79.348773    4.0           116        1   
74  43.765266 -79.166977    1.5             3        1   

                                           categories  
9   martial arts, gyms, fitness & instruction, act...  
26                     shopping, shoe stores, fashion  
52  cosmetics & beauty supply, shopping, beauty & ...  
73                            vietnamese, restaurants  
74  event planning & services, photo booth rentals...

In [122]:
df2 = df1.groupby(['name'])
df2.head(4)

name  \
9                                          xtreme couture   
26                                       air jordan store   
52                                    bath and body works   
73                                       mi mi restaurant   
74      equipment sales and long term rentals - redlin...   
...                                                   ...   
209330                                unity auto shipping   
209351                                     the black bull   
209359                                       q's shawarma   
209377             steak & cheese & quick pita restaurant   
209392                                          kudlow ye   

                                     address     city state postal_code  \
9               700 kipling avenue etobicoke  toronto    on     M8Z 5G3   
26                          306 yonge street  toronto    on     M5B 1R4   
52                      1900 eglintion ave e  toronto    on     M1L 2L9   
73                      688 gerrard street e  toronto    on     M4M 1Y3   
74      510 coronation dr. unit #18, unit 18  toronto    on     M1E 4X6   
...                                      ...      ...   ...         ...   
209330     2937 eglinton avenue e, suite 100  toronto    on     M1J 2E5   
209351                    298 queen street w  toronto    on     M5V 2A1   
209359                1075 martin grove road  toronto    on     M9W 4W6   
209377                        199 college st  toronto    on     M5T 1P9   
209392       21 st clair avenue e, suite 401  toronto    on     M4T 1L9   

         latitude  longitude  stars  review_count  is_open  \
9       43.624539 -79.529108    3.0            16        1   
26      43.656542 -79.381308    4.0             9        1   
52      43.727189 -79.293008    3.5             7        1   
73      43.666376 -79.348773    4.0           116        1   
74      43.765266 -79.166977    1.5             3        1   
...           ...        ...    ...           ...      ...   
209330  43.739323 -79.234861    5.0             3        1   
209351  43.649634 -79.393307    2.5           153        1   
209359  43.701807 -79.575135    4.0            16        1   
209377  43.658683 -79.395071    2.5            21        1   
209392  43.688214 -79.392756    5.0             3        1   

                                               categories  
9       martial arts, gyms, fitness & instruction, act...  
26                         shopping, shoe stores, fashion  
52      cosmetics & beauty supply, shopping, beauty & ...  
73                                vietnamese, restaurants  
74      event planning & services, photo booth rentals...  
...                                                   ...  
209330  automotive, car dealers, towing, vehicle shipping  
209351                              pubs, bars, nightlife  
209359                         restaurants, mediterranean  
209377                            sandwiches, restaurants  
209392  tax services, professional services, accountan...  

[13784 rows x 11 columns]

In [123]:
df1

name  \
9                                          xtreme couture   
26                                       air jordan store   
52                                    bath and body works   
73                                       mi mi restaurant   
74      equipment sales and long term rentals - redlin...   
...                                                   ...   
209351                                     the black bull   
209359                                       q's shawarma   
209374                                          starbucks   
209377             steak & cheese & quick pita restaurant   
209392                                          kudlow ye   

                                     address     city state postal_code  \
9               700 kipling avenue etobicoke  toronto    on     M8Z 5G3   
26                          306 yonge street  toronto    on     M5B 1R4   
52                      1900 eglintion ave e  toronto    on     M1L 2L9   
73                      688 gerrard street e  toronto    on     M4M 1Y3   
74      510 coronation dr. unit #18, unit 18  toronto    on     M1E 4X6   
...                                      ...      ...   ...         ...   
209351                    298 queen street w  toronto    on     M5V 2A1   
209359                1075 martin grove road  toronto    on     M9W 4W6   
209374                     81 front street e  toronto    on     M5E 1B8   
209377                        199 college st  toronto    on     M5T 1P9   
209392       21 st clair avenue e, suite 401  toronto    on     M4T 1L9   

         latitude  longitude  stars  review_count  is_open  \
9       43.624539 -79.529108    3.0            16        1   
26      43.656542 -79.381308    4.0             9        1   
52      43.727189 -79.293008    3.5             7        1   
73      43.666376 -79.348773    4.0           116        1   
74      43.765266 -79.166977    1.5             3        1   
...           ...        ...    ...           ...      ...   
209351  43.649634 -79.393307    2.5           153        1   
209359  43.701807 -79.575135    4.0            16        1   
209374  43.648727 -79.372506    4.0            22        1   
209377  43.658683 -79.395071    2.5            21        1   
209392  43.688214 -79.392756    5.0             3        1   

                                               categories  
9       martial arts, gyms, fitness & instruction, act...  
26                         shopping, shoe stores, fashion  
52      cosmetics & beauty supply, shopping, beauty & ...  
73                                vietnamese, restaurants  
74      event planning & services, photo booth rentals...  
...                                                   ...  
209351                              pubs, bars, nightlife  
209359                         restaurants, mediterranean  
209374                                 coffee & tea, food  
209377                            sandwiches, restaurants  
209392  tax services, professional services, accountan...  

[14928 rows x 11 columns]

In [129]:
df1.to_csv("cleaned_data.csv", index=False)
#df2.to_csv("separated_categories.csv", index=False)